File: src/leetcode/p121_best_time_to_buy_and_sell_stock.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # C1: max profit -> max(res,(D[Sell] - D[Buy]))
        # C2: no profit -> return 0
        # To acheive profit, we are looking for upward slope
        # How do we exhaust all possibility?

        # Solution 1: double for loop -> O(n^2)
        # Solution 2: sliding window
            # when do we move the left and right pointers?
            # we want lowest buy price -> have lower price, move the that point
            # we want highest buy price -> have higher price, move the that point

        res = 0

        if len(prices)<=1:
            return res

        # assume buying in the 0th day and sell in the 1th day
        l,r = 0,1

        while l<r:
            profit = prices[r] - prices[l]
            res = max(res,profit)
            if profit<0:
                l = r
                if r != len(prices)-1:
                    r+=1
            else:
                if r != len(prices)-1:
                    r+=1
                else:
                    l+=1


        # return 0 if no profit
        return res

File: src/leetcode/p121_best_time_to_buy_and_sell_stock.rs
#![allow(dead_code)]

use super::Solution;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        // base case, return 0 if no profit
        let mut res = 0;

        if prices.len() <= 1 {
            return res;
        }

        let (mut l, mut r) = (0, 1);

        while l < r {
            let profit = prices[r] - prices[l];
            res = std::cmp::max(res, profit);

            if profit < 0 {
                l = r;
                if r != prices.len() - 1 {
                    r += 1;
                }
            } else {
                if r != prices.len() - 1 {
                    r += 1
                } else {
                    l += 1
                }
            }
        }

        res
    }
}

#[cfg(test)]
#[test]
fn main() {
    assert_eq!(Solution::max_profit(vec![7, 1, 5, 3, 6, 4]), 5);
    assert_eq!(Solution::max_profit(vec![7, 6, 4, 3, 1]), 0)
}
